xtext/org.icyphy.linguafranca/src/org/icyphy/generator/GeneratorBase.xtend

```diff
@@ -603,6 +632,205 @@ abstract class GeneratorBase extends AbstractLinguaFrancaValidator {
+                        }
+                    }
+                    count++;
+                }
+            }
+            // Next, set up the downstream array.
+            if (!federate.sendsTo.keySet.isEmpty) {
+                // Federate sends non-physical messages to other federates.
+                // Initialize the downstream array.
+                val numDownstream = federate.sendsTo.keySet.size
+                // Allocate memory for the array.
+                pr(rtiCode, '''
+                    federates[«federate.id»].downstream = malloc(sizeof(federate_t*) * «numDownstream»);
+                    federates[«federate.id»].num_downstream = «numDownstream»;
+                ''')
+                // Next, populate the array.
+                // Find the minimum delay in the process.
+                // FIXME: Zero delay is not really the same as a microstep delay.
+                var count = 0;
+                for (downstreamFederate : federate.sendsTo.keySet) {
+                    pr(rtiCode, '''
+                        federates[«federate.id»].downstream[«count»] = «downstreamFederate.id»;
+                    ''')
+                    count++;
+                }
+            }
+        }
+        
+        // Start the RTI server before launching the federates because if it
+        // fails, e.g. because the port is not available, then we don't want to
+        // launch the federates.
+        pr(rtiCode, '''
+            int socket_descriptor = start_rti_server(«federationRTIProperties.get('port')»);
+        ''')
+        
+        // Generate code that blocks until the federates resign.
+        pr(rtiCode, "wait_for_federates(socket_descriptor);")
+        
+        unindent(rtiCode)
+        pr(rtiCode, "}")
+        
+        var fOut = new FileOutputStream(
+                new File(directory + rtiSrcPath + File.separator + cFilename));
+        fOut.write(rtiCode.toString().getBytes())
+        fOut.close()
+    }
+    
+    
+        /** Invoke the compiler on the generated code. */
+    def compileCode() {
+
+        // If there is more than one federate, compile each one.
+        var fileToCompile = filename // base file name.
+        for (federate : federates) {
+            // Empty string means no federates were defined, so we only
+            // compile one file.
+            if (!federate.isSingleton) {
+                fileToCompile = filename + '_' + federate.name
+            }
+            executeCommand(compileCommand(fileToCompile, true), directory)
+        }
+        // Also compile the RTI files if there is more than one federate.
+        if (federates.length > 1) {
+            compileRTI()
+        }
+    }
+    
+    /** Invoke the compiler on the generated RTI */
+    def compileRTI() {
+        var fileToCompile = filename + '_RTI'
+        executeCommand(compileCommand(fileToCompile, false), directory)
+    }
+    
+    /** Return a command to compile the specified C file.
```

Comment:
Also C-specific. Doesn't belong.

Reply:
The intent of these is that all targets should be using the same RTI. So either you have to design your own (likely incompatible) RTI, or need to adopt a common one.  How do you suggest dealing with that?  Should the TS code generator just generate incomplete code, with no RTI?  How would you generate the RTI?

Reply:
I'm not questioning the intent, I'm questioning one very specific aspect of the implementation: the leaking of target-specific code into a generic base class and into code generators with different targets.

Reply:
If the same C RTI is used for all targets, doesn't that mean the functions used to generate the RTI aren't target-specific? As far as I see, there's no leaking going on here.

Reply:
I don't know what else I can do you illuminate the problem other than what I've already stated... The file is called `GeneratorBase`, not `RTIGenerator`...

Reply:
I agree it would be an improvement to separate the RTI specific code from GeneratorBase (and I'll start looking into that), but I maintain that GeneratorBase should provide all target language generators with the capability to produce an RTI. Target language generators need to do the following things:

- Determine the location where the RTI code and binary should be placed 
- Compile the RTI at the appropriate time
- Report errors about the RTI generation along with other compilation errors

Since this is shared among all target language generators, GeneratorBase should provide these capabilities. But maybe GeneratorBase can pass off the implementation of these functions to an RTIGenerator (right now there is only a CRTIGenerator but there could be more and RTIGenerator would be an abstract class).

Reply:
I never suggested the GeneratorBase shouldn't offer the capabilities that you enumerate. All I'm saying is that GeneratorBase shouldn't have any C (or any other target) code or compiler configuration in it. Yes, this requires some careful thought, but that's necessary for the sake of keeping a clean and maintainable code base.